OpenSim/Tests/Wrapping/testWrapCylinder.cpp

```diff
@@ -0,0 +1,314 @@
+
+        std::unique_ptr<WrapCylinder> cylinder(new WrapCylinder());
+        cylinder->setName("cylinder");
+        cylinder->set_radius(input.radius);
+        cylinder->set_length(input.cylinderLength);
+        cylinder->set_quadrant(input.quadrant);
+        cylinder->set_xyz_body_rotation(input.eulerRotations);
+        cylinder->setFrame(model.getGround());
+
+        std::unique_ptr<PathSpring> spring (new PathSpring("spring", 1., 1., 1.));
+        spring->updGeometryPath().appendNewPathPoint(
+            "startPoint",
+            model.get_ground(),
+            input.path.start);
+        spring->updGeometryPath().appendNewPathPoint(
+            "endPoint",
+            model.get_ground(),
+            input.path.end);
+        spring->updGeometryPath().addPathWrap(*cylinder.get());
+
+        model.addComponent(spring.release());
+        model.updGround().addWrapObject(cylinder.release());
+
+        model.finalizeConnections();
+        model.setUseVisualizer(visualize);
+
+        const SimTK::State& state = model.initSystem();
+
+        if (visualize) {
+            model.realizeVelocity(state);
+            model.getVisualizer().show(state);
+        }
+
+        model.updComponent<PathSpring>("spring").getLength(state);
+        WrapResult wrapResult = model.updComponent<PathSpring>("spring")
+                .getGeometryPath()
+                .getWrapSet()
+                .get("pathwrap")
+                .getPreviousWrap();
+
+        // Convert the WrapResult to a WrapTestResult for convenient assertions.
+        WrapTestResult result;
+        result.path = PathSegment{
+            wrapResult.r1,
+            wrapResult.r2,
+        };
+        result.length = wrapResult.wrap_path_length;
+        result.noWrap = wrapResult.wrap_pts.size() == 0;
+
+        return result;
+    }
+
+    // Simulates and tests a wrapping case:
+    // - Computes the wrapping result given the input.
+    // - Tests if computed result is equal to the expected result.
+    // - Returns a string containing info on the failed test, or an empty string
+    // if succesful.
+    std::string TestWrapping(
+        const WrapInput& input,
+        const WrapTestResult& expected,
+        WrappingTolerances tol,
+        std::string testCase,
+        bool visualize = false)
+    {
+
+        WrapTestResult result = solve(input, visualize);
+
+        if (!IsEqual(result, expected, tol.position)) {
+            std::ostringstream oss;
+            oss << "\nFAILED: case = " << testCase;
+            oss << "\n    input = " << input;
+            oss << "\n    result = " << result;
+            oss << "\n    expected = " << expected;
+            oss << "\n    caused by: "
+                << "Expected and simulated result not equal to within tolerance = "
+                << tol.position;
+            return oss.str();
+        }
+
+        return std::string{};
+    }
+
+}
+
+int main()
+{
+    std::string failLog;
```

Comment:
`std::vector<std::string>`?

Reply:
Yes, because it makes it explicit in the source code that the variable is "a sequence of error messages" rather than relying on an in-string encoding.